Refactor/coins endpoints

agusbusta opened the pull request:
Create CoinBot (/coins)
This POST endpoint allows creating a new CoinBot in the system. It handles the upload of SVG icons to AWS S3 and the association with an existing category. It performs input data validations and manages database and file processing errors.
Update CoinBot (/coins/int:coin_id)
This PUT endpoint updates an existing CoinBot. It allows modifying fields such as name, alias, category, background color, and icon. It handles icon updates in AWS S3, including the deletion of the old icon if necessary. It implements robust validations and error handling.
Delete CoinBot (/coins/int:coin_id)
This DELETE endpoint removes a specific CoinBot and all its associated data. It deletes the icon from AWS S3 if it exists. It uses cascade deletion to ensure all related data is properly removed. It includes error handling for cases where the CoinBot is not found or database failures occur.
Activate/Deactivate CoinBot (/coins/int:coin_id/toggle-publication)
This POST endpoint manages the activation and deactivation of a CoinBot. When activating, it performs validations of the CoinBot's fundamentals and chart. If the validations pass, it activates the CoinBot; if not, it provides detailed error messages. Deactivation is straightforward without additional validations. It updates the 'is_active' field and records the action time.

daviddflix commented on routes/coin_bot/coinbot.py:
```diff
@@ -1,88 +1,294 @@
+                return jsonify(response), 404
+
+            # Update fields if provided
+            for field in ['name', 'alias', 'category_id', 'background_color', 'is_active']:
```
- The field is active is not required, False should be the default state when a coin is created, this should be active only when its fundamentals and chart have data.

daviddflix commented on routes/coin_bot/coinbot.py:
```diff
@@ -1,88 +1,294 @@
+
+            session.commit()
+            response["success"] = True
+            response["coin_bot"] = coin_bot.as_dict()
```
- Update the name of the variable.

daviddflix commented on routes/coin_bot/coinbot.py:
```diff
@@ -1,88 +1,294 @@
+
+
+
+@coin_bot_bp.route('/coins/<int:coin_id>', methods=['DELETE'])
```
- Rename the endpoints as this should be in singular.

daviddflix commented on routes/coin_bot/coinbot.py:
```diff
@@ -1,88 +1,294 @@
+                    image_processor.delete_from_s3(S3_BUCKET_ICONS, icon_filename)
+                except Exception as e:
+                    # Log the error, but continue with deletion
+                    print(f"Error deleting icon from S3: {str(e)}")
```
- Remove all print statements and if it raises an error, this should be return.

daviddflix reviewed (changes requested):
## Feedback

@agusbusta, I've reviewed the endpoints are in general they have a good structure and they are good, I have highlighted some small changes.

- Init the imageprocessor without the env variable, the class doesn't need them anymore.

Addition of new endpoint:
- Please create a new endpoint, this should be the official endpoint to retrieve all coins, the name of this new endpoint should be "/coins" in plural. and it should get all coins or return an empty array.

agusbusta commented:
Summary of Changes:

1. Error Handling Improvement in delete_coin Function:
   - Removed print statement for S3 icon deletion errors
   - Now returns an error response if S3 icon deletion fails, preventing inconsistent states

2. New Endpoint for Retrieving All Coins:
   - Added GET endpoint '/coins' to coin_bp Blueprint
   - Retrieves all CoinBot instances from the database
   - Returns JSON with success status, coin array, and error information
   - Implements error handling for database and unexpected errors
   - Uses appropriate HTTP status codes (200 for success, 500 for errors)

3. Code Cleanup and Best Practices:
   - Removed unnecessary print statements throughout the code
   - Improved error handling to return errors instead of logging and continuing

4. Consistency in Response Formats:
   - Standardized JSON response structure across endpoints
   - All endpoints now return a consistent format with 'success', relevant data, and 'error' fields

agusbusta commented:
Improvements to the PUT route for coin updates

This pull request includes the following enhancements to the `update_coin` function:

1. Added docstrings:
   - Detailed documentation has been added to the `update_coin` function, including a general description, parameters, return value, and possible exceptions.
   - This improves code readability and understanding for other developers.

2. Modification in S3 icon deletion:
   - Updated the logic for deleting the old icon from S3 AWS before uploading a new one.
   - Now using the `delete_from_s3` function from the `image_processor` to ensure proper file deletion.

3. Alias normalization for icon filenames:
   - Implemented normalization of the coin's alias when generating the filename for the new icon.
   - The alias is now converted to lowercase, whitespace is removed, and a consistent format for icon filenames is ensured.

daviddflix commented:
## Feedback

Path: routes/category/category.py
Remove env variables from image_processor.


Line 63: `icon_filename = secure_filename(f"{alias}.svg")`
Normalize the alias:
- Convert to lowercase.
- Remove spaces from the beginning, middle, and end of the string.

### delete_category endpoint:
- Ensure proper session handling at line 120: session.begin_nested(). Make sure the session is properly closed to avoid potential issues.
The coins associated with the category are currently being set to null instead of being deleted, which can lead to orphaned coins and database conflicts. Adjust the code to delete all associated data for each coin in the category. This includes deleting:
Fundamentals
Charts
Coin data
update endpoint:

Line 226: image_processor.delete_from_s3(S3_BUCKET_ICONS, old_icon_filename)
The delete method should take the icon URL.
Important: Ensure the method handles SVG file deletion, as it seems to be currently designed for PNG or JPEG files only.
Line 228: new_icon_filename = secure_filename(f"{category.alias or category.name}_icon.svg")
Normalize the alias:
Convert to lowercase.
Remove spaces and replace with hyphens.
Use the alias only for the new icon filename.
The session in this endpoint is not managed correctly. It is using the global session, which can lead to conflicts. Adjust to properly manage the session within the endpoint.
@category_bp.route('/coins/global-toggle', methods=['POST']):

Rename this route to: /categories/global-toggle.
toggle_category_coins endpoint:

This endpoint does not manage the session correctly. Ensure the session is initialized and closed within the endpoint.
global_toggle_coins endpoint:

Similarly, this endpoint also does not handle the session properly. Adjust the session management.
Utility function: validate_coin_bot:

This function is using the global session. Modify it to open and close a session within the function.
Important
Please record a video demonstrating the manual tests you performed after all the improvements. Ensure that all endpoints are thoroughly tested again.

agusbusta commented:
@daviddflix get route with alphabetic order was added.

daviddflix commented:
# Feedback

Endpoint update:
- Symbol was not being updated.
- update icon is not being properly done, as it is taken the new icon as a string in the json, this not the correct approach as the user could enter whatever string and break all.

Create endpoint
- symbol was being received by files, which is incorrect.
